src/qat/purr/compiler/instructions.py

```diff
@@ -741,3 +752,153 @@ def calculate_duration(instruction, return_samples: bool = True):
+        self._duration = 0.0
+        self.add_measurements(targets, mode, output_variables, entangled_qubits)
+
+    def add_measurements(
+        self,
+        targets: Union[Qubit, List[Qubit]],
+        mode: AcquireMode,
+        output_variables: Union[str, List[str]] = None,
+        entangled_qubits: List[Qubit] = None,
+        existing_names: Set[str] = None,
+    ):
+        targets = self._validate_targets(targets, (Qubit))
+        if len((duplicates := [t for t in targets if t in self.quantum_targets])) > 0:
+            raise ValueError(
+                "Target can only be measured once in a 'MeasureBlock'. "
+                f"Duplicates: {duplicates}"
+            )
+        if not isinstance(output_variables, list):
+            output_variables = [] if output_variables is None else [output_variables]
+        if (num_out_vars := len(output_variables)) == 0:
+            output_variables = [None] * len(targets)
+        elif num_out_vars != len(targets):
+            raise ValueError(
+                f"Unsupported number of `output_variables`: {num_out_vars}, "
+                f"must be `None` or match numer of targets: {len(targets)}."
+            )
+        for target, output_variable in zip(targets, output_variables):
+            meas, acq = self._generate_measure_acquire(
+                target, mode, output_variable, existing_names
+            )
+            self._duration = max(self._duration, meas.duration, acq.delay + acq.duration)
+            self._target_dict[target.full_id()] = {
+                "target": target,
+                "mode": mode,
+                "output_variable": output_variable,
+                "measure": meas,
+                "acquire": acq,
+            }
+        self._entangled_qubits.update(
+            self._validate_targets(entangled_qubits, (Qubit))
+            if entangled_qubits is not None
+            else targets
+        )
+
+    @property
+    def quantum_targets(self):
+        return [v["target"] for v in self._target_dict.values()]
+
+    def get_acquires(self, targets: Union[Qubit, List[Qubit]]):
+        if not isinstance(targets, list):
+            targets = [targets]
+        return [self._target_dict[qt.full_id()]["acquire"] for qt in targets]
+
+    def __repr__(self):
+        target_strings = []
+        for q, d in self._target_dict.items():
+            out_var = (
+                f"->{d['output_variable']}" if d["output_variable"] is not None else ""
+            )
+            mode = f":{d['mode'].value}" if d["mode"] is not None else ""
+            target_strings.append(f"{q}{mode}{out_var}")
+        return f"Measure {', '.join(target_strings)}"
+
+    @property
+    def instructions(self) -> List[QuantumInstruction]:
+        instructions = [Synchronize(list(self._entangled_qubits))]
+        for _, values in self._target_dict.items():
+            instructions.extend([values["measure"], values["acquire"]])
```

Comment:
So just to confirm for clarity (I didn't realise this until now) measure stacks the measures of different targets sequentially but in different channels so when breaking down the timings later. All sequential measurements happen simultaneously?  Does this then mean if you do measure Q1 sync and measure Q2 they actually happen together NOT sequentially? @bryn-bell

Reply:
That's good! I think I'm still not understanding this, what is stopping it? Doesn't a sync count as an optional measure instr and get pulled into a measure block? And then the pervious qubit measurement is also pulled into the measure block?

Reply:
I've added a couple of tests to check this for the new `MeasureBlock` version.